typed/declaration/schema-schema.ipldsch

```diff
@@ -0,0 +1,358 @@
+## -----
+## This is the schema describing the schema declarations for IPLD Schemas.
+## Yes, it's self-describing! :)
+## -----
+
+## Type names are a simple alias of string.
+##
+## There are some additional rules that should be applied:
+##   - Type names should by convention begin with a capital letter;
+##   - Type names must be all printable characters (no whitespace);
+##   - Type names must not contain punctuation (dashes, dots, etc).
+##
+## Type names are strings meant for human consumption at a local scope.
+## When making a Schema, note that the TypeName is the key of the map:
+## a TypeName must be unique within the Schema.
+##
+type TypeName string
+
+## SchemaMap is a complete set of types;
+## it is simply a map of TypeName to detailed declaration of that Type.
+##
+## A simple schema map with one type might look like this:
+##
+## ```
+## {
+##   "MyFooType": {
+##     "type": "string"
+##   }
+## }
+## ```
+##
+type SchemaMap map {TypeName:Type}
+
+## Schema is a single-member union, which can be used in serialization
+## to make a form of "nominative type declaration".
+##
+## A complete (if quite short) Schema might look like this:
+##
+## ```
+## {
+##   "schema": {
+##     "MyFooType": {
+##       "type": "string"
+##     }
+##   }
+## }
+## ```
+##
+type Schema union {
+	| SchemaMap "schema"
```

Comment:
Yup.  Pipe is a visual cue that we're in a union or enum.

(A lot of the syntactical choices thus far are fairly arbitrary; the parse could even be unambiguous without this.  It's more or less there because I liked the drafts better that made sum types (union, enum) look strongly visually distinct.  And probably, subconsciously, the Pipe character is coming from either Haskell, [OCaml variants](https://v1.realworldocaml.org/v1/en/html/variants.html), or [GraphQL unions](https://graphql.org/learn/schema/#union-types).)